lib/rpc/singleJsonRpcProvider.ts

```diff
@@ -0,0 +1,460 @@
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+import { Network } from '@ethersproject/networks'
+import { Deferrable } from '@ethersproject/properties'
+const debug = Debug('SingleJsonRpcProvider')
+
+class PerfStat {
+  lastCallTimestampInMs: number = 0
+  lastCallSucceed: boolean = false
+  lastCallLatencyInMs: number = 0
+  timeWaitedBeforeLastCallInMs: number = 0
+}
+
+export default class SingleJsonRpcProvider extends StaticJsonRpcProvider {
+  // TODO(jie): Implement block-aligned cache
+  url: string
+  private healthScore
+  private isRecovering: boolean
+  private perf: PerfStat
+  private config: Config
+  private readonly metricPrefix: string
+
+  constructor(chainId: ChainId, url: string, config: Config = DEFAULT_CONFIG) {
+    super(url, { chainId, name: CHAIN_IDS_TO_NAMES[chainId] })
+    this.url = url
+    this.healthScore = 0
+    this.isRecovering = false
+    this.perf = new PerfStat()
+    this.config = config
+    this.metricPrefix = `RPC_${this.network.chainId}_${this.url}`
+  }
+
+  isHealthy(): boolean {
+    return !this.isRecovering
+  }
+
+  hasEnoughWaitSinceLastCall(): boolean {
+    debug(`${this.url}: score ${this.healthScore}, waited ${Date.now() - this.perf.lastCallTimestampInMs} ms`)
+    return Date.now() - this.perf.lastCallTimestampInMs > this.config.RECOVER_EVALUATION_WAIT_PERIOD_IN_MS
+  }
+
+  private recordError() {
+    this.healthScore += this.config.ERROR_PENALTY
+    debug(`${this.url}: error penalty ${this.config.ERROR_PENALTY}, score => ${this.healthScore}`)
+  }
+
+  private recordHighLatency() {
+    this.healthScore += this.config.HIGH_LATENCY_PENALTY
+    debug(`${this.url}: high latency penalty ${this.config.ERROR_PENALTY}, score => ${this.healthScore}`)
+  }
+
+  private recordProviderRecovery(timeInMs: number) {
+    if (this.healthScore === 0) {
+      return
+    }
+    this.healthScore += timeInMs * this.config.RECOVER_SCORE_PER_MS
+    if (this.healthScore > 0) {
+      this.healthScore = 0
+    }
+    debug(
+      `${this.url}: healthy: ${this.isHealthy()}, recovery ${timeInMs} * ${this.config.RECOVER_SCORE_PER_MS} = ${
+        timeInMs * this.config.RECOVER_SCORE_PER_MS
+      }, score => ${this.healthScore}`
+    )
+  }
+
+  private checkLastCallPerformance(method: string) {
+    if (!this.perf.lastCallSucceed) {
+      metric.putMetric(`${this.metricPrefix}_${method}_FAILED`, 1, MetricLoggerUnit.Count)
+      this.recordError()
+    } else if (this.perf.lastCallLatencyInMs > this.config.MAX_LATENCY_ALLOWED_IN_MS) {
+      metric.putMetric(`${this.metricPrefix}_${method}_SUCCESS_HIGH_LATENCY`, 1, MetricLoggerUnit.Count)
+      this.recordHighLatency()
+    } else {
+      metric.putMetric(`${this.metricPrefix}_${method}_SUCCESS`, 1, MetricLoggerUnit.Count)
+      // For a success call, we will increase health score.
+      if (this.perf.timeWaitedBeforeLastCallInMs > 0) {
+        this.recordProviderRecovery(this.perf.timeWaitedBeforeLastCallInMs)
```

Comment:
so we increase proportionally to the time between the last call? whats the intuition behind this?

Reply:
The longer we wait the more likely for it to have recovered, given that the last call was succesful?